test/balancer_test.go

```diff
@@ -981,3 +981,266 @@ func (s) TestSubConnShutdown(t *testing.T) {
+			ccw.stateListener(scs)
+			origListener(scs)
+		}
+	}
+	sc, err := ccw.ClientConn.NewSubConn(addrs, opts)
+	ccw.scChan <- sc
+	return sc, err
+}
+
+// Test calls RegisterHealthListener on a SubConn to verify that expected health
+// updates are sent only to the most recently registered listener.
+func (s) TestSubConn_RegisterHealthListener(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	scChan := make(chan balancer.SubConn, 1)
+	bf := stub.BalancerFuncs{
+		Init: func(bd *stub.BalancerData) {
+			cc := bd.ClientConn
+			ccw := &subConnStoringCCWrapper{
+				ClientConn: cc,
+				scChan:     scChan,
+			}
+			bd.Data = balancer.Get(pickfirst.Name).Build(ccw, bd.BuildOptions)
+		},
+		Close: func(bd *stub.BalancerData) {
+			bd.Data.(balancer.Balancer).Close()
+		},
+		UpdateClientConnState: func(bd *stub.BalancerData, ccs balancer.ClientConnState) error {
+			return bd.Data.(balancer.Balancer).UpdateClientConnState(ccs)
+		},
+		ExitIdle: func(bd *stub.BalancerData) {
+			bd.Data.(balancer.ExitIdler).ExitIdle()
+		},
+	}
+
+	stub.Register(t.Name(), bf)
+	svcCfg := fmt.Sprintf(`{ "loadBalancingConfig": [{%q: {}}] }`, t.Name())
+	backend := stubserver.StartTestService(t, nil)
+	defer backend.Stop()
+	opts := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithDefaultServiceConfig(svcCfg),
+	}
+	cc, err := grpc.NewClient(backend.Address, opts...)
+	if err != nil {
+		t.Fatalf("grpc.NewClient(%q) failed: %v", backend.Address, err)
+
+	}
+	defer cc.Close()
+
+	cc.Connect()
+
+	var sc balancer.SubConn
+	select {
+	case sc = <-scChan:
+	case <-ctx.Done():
+		t.Fatal("Context timed out waiting for SubConn creation")
+	}
+	healthUpdateChan := make(chan balancer.SubConnState, 1)
+
+	// Register listener while Ready and verify it gets a health update.
+	testutils.AwaitState(ctx, t, cc, connectivity.Ready)
+	for i := 0; i < 2; i++ {
+		sc.RegisterHealthListener(func(scs balancer.SubConnState) {
+			healthUpdateChan <- scs
+		})
+		select {
+		case scs := <-healthUpdateChan:
+			if scs.ConnectivityState != connectivity.Ready {
+				t.Fatalf("Received health update = %v, want = %v", scs.ConnectivityState, connectivity.Ready)
+			}
+		case <-ctx.Done():
+			t.Fatalf("Context timed out waiting for health update")
+		}
+
+		// No further updates are expected.
+		select {
+		case scs := <-healthUpdateChan:
+			t.Fatalf("Received unexpected health update while channel is in state %v: %v", cc.GetState(), scs)
+		case <-time.After(defaultTestShortTimeout):
+		}
+	}
+
+	// Make the SubConn enter IDLE and verify that health updates are recevied
+	// on registering a new listener.
+	backend.S.Stop()
+	backend.S = nil
+	testutils.AwaitState(ctx, t, cc, connectivity.Idle)
+	if err := backend.StartServer(); err != nil {
+		t.Fatalf("Error while restarting the backend server: %v", err)
+	}
+	cc.Connect()
+	testutils.AwaitState(ctx, t, cc, connectivity.Ready)
```

Comment:
What is the point entering IDLE here since we register the health listener only after entering READY again?

Reply:
This simulates how LB policy will behave. When the subchannel becomes IDLE, they will call connect and wait for the subchannel to become ready and then register a health listener. It verifies that the subchannel correctly handles this scenario.

Looking at the implementation, it may look like the code flow is the same irrespective of the the number of times the subchannel has entered READY, but the implementation may change and someone may make the mistake of assuming that subchannels enter READY only once.